src/register-service-worker.js

```diff
@@ -0,0 +1,162 @@
+/* eslint-env browser */
+
+(function() {
+  var constants = require('./lib/constants.js');
+
+  var swScript = document.currentScript.dataset.serviceWorker;
+  var manifestAttribute = document.documentElement.getAttribute('manifest');
+
+  if (swScript && manifestAttribute && 'serviceWorker' in navigator) {
+    var manifestUrl = (new URL(manifestAttribute, location.href)).href;
+
+    var idb = require('idb');
+
+    idb.open(constants.DB_NAME, constants.DB_VERSION, function(upgradeDB) {
+      if (upgradeDB.oldVersion === 0) {
+        Object.keys(constants.OBJECT_STORES).forEach(function(objectStore) {
+          upgradeDB.createObjectStore(constants.OBJECT_STORES[objectStore]);
+        });
+      }
+    }).then(function(db) {
+      return Promise.all([
+        updateManifestAssociationForCurrentPage(db, manifestUrl),
+        handlePossibleManifestUpdate(db, manifestUrl)
+      ]);
+    }).then(function() {
+      return navigator.serviceWorker.register(swScript);
+    });
+  }
+
+  /**
+   * Caches the Responses for one or more URLs, using the Cache Storage API.
+   *
+   * @param {Array.<String>} urls
+   * @returns {Promise.<T>}
+   */
+  function addToCache(urls) {
+    return caches.open(constants.CACHE_NAME).then(function(cache) {
```

Comment:
I wasn't taking manifest-to-cache scoping into account at all, let alone the additional manifest-version-to-cache scoping behavior you describe. Thanks for pointing that out.

I'd like to match the AppCache behavior when at all possible, but am not sure how that `postMessage` communication would work. Here's what comes to mind, but please clarify if I'm misinterpreting:

- Page A is open, using manifest Z with version 1. (In practice, I think I'm going to have to hash the manifest contents to assign them version ids, but let's just say 1.)
- Page B is opened in a new tab, using manifest Z, and a manifest update is detected, leading to version 2.
- At this point, page B sends a `postMessage` to the SW that it's now on version 2 of manifest Z. (I think I'd have to communicate this using the SW as an intermediary, since I'm not sure how to `postMessage` from Page B to Page A.)
- The SW sends a `postMessage` to everything in `self.clients` that it can identify which are using manifest Z version 1.
- Page A gets that message from the SW, and... sets up an `onunload` handler to clear out the old cache? Something else? I can't detect from the SW when a client goes away, right? And that logic would get messier if there is more than one tab open using manifest Z version 1.